README.txt
UbuntuMan README

Copyright (C) 2008 Henri Häkkinen

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program.  If not, see <http://www.gnu.org/licenses/>.


To download the latest version of this software, please visit:

    http://henux.nor.fi/coding/projects/9-other/7-ubuntuman
    http://github.com/henux/ubuntuman/tree/master


INTRODUCTION

UbuntuMan is a Supybot plugin for browsing online manual pages from the
Ubuntu Manpage Repository (http://manpages.ubuntu.com) in an IRC channel.

The plugin arose from a need to have a fancy manual page browser in
##club-ubuntu IRC channel at irc.freenode.net server.  Several versions of
the plugin was made by the same author, until the Ubuntu Manpage Repository
was decided on.

Thanks to emma and bascule for their continued support and several others
in ##club-ubuntu for bug testing.  Also thanks to Terence Simpson for his
contribution and to Elián Hanisch for non-English manual page support.


USAGE

The plugin provides the following new Supybot commands:

man <command> [--rel <release>] [--lang <language>]

    Prints a short version of the manual page from the Ubuntu Manpage
    Repository. The format is defined in UbuntuMan.format configuration
    variable.

manurl <command> [--rel <release>] [--lang <language>]

    Gives the URL of the full manual page in the Ubuntu Manpage Repository.

Commands accept --rel and --lang options, which can be used to override the
default Ubuntu release and language the manual pages are fetched for.  For
example, to see the Spanish manual page for the 'ls' command as it exists
in Ubuntu Hardy , use "man ls --rel hardy --lang es".  If these options are
not given, the configured defaults are used.


CONFIGURATION VARIABLES

supybot.plugins.UbuntuMan.baseurl

    The base URL to the Canonical's Ubuntu Manpage Repository.  In all
    normal cases, the default value should be used.  This variable should
    never end in slash!
    Default value: http://manpages.ubuntu.com/manpages

supybot.plugins.UbuntuMan.release

    The default release to fetch the manual pages for.  This should be one
    of the following: dapper, gutsy, hardy, intrepid, jaunty.
    Default value: intrepid

supybot.plugins.UbuntuMan.sections

    Space separated list of enabled sections.  Note this variable has a
    different syntax than in the version 1.0.
    Default value: 1 5 8

supybot.plugins.UbuntuMan.language

    The default language used to fetch the manual pages for.  This should
    be one of the following: de, en, es, fi, fr, it.
    Default value: en

supybot.plugins.UbuntuMan.format

    The format of the 'man' command. This should be a string containing any
    of the following keywords:
    $name - manpage NAME section.
    $synopsis - manpage SYNOPSIS section.
    $description - manpage DESCRIPTION section.
    $command - the name of the command.
    $url - the url address where the manpage is located.
    Default value:
    $name | $synopsis | $description

supybot.plugins.UbuntuMan.maxLength

    Maximun number of characters the 'man' command can use. If the output
    excess this value one of the manpage sections (normally DESCRIPTION) is
    cut for fit the limit, if still is too long then the whole reply is cut.
    Note that supybot.reply.mores.length takes priority over this variable if
    its value is other than zero.
    Default value: 300



NEWS
UbuntuMan 1.2.0
===============

New features and bugfixes

    - Support for formatting man's output.
    - Support for Italian and French manual pages.
    - Fix: apt-get Spanish manpage failed parsing.
    - Fix: in some cases, UbuntuMan would try the same
    url twice.
    - Fix: supybot would raise an exception if couldn't
    find a section in a non-English manpage.


UbuntuMan 1.1.5
===============

Minor changes

    - Update for the Canonical's manpage repository changes.


UbuntuMan 1.1.4
===============

New features and changes

    - Support for non-English manual pages.
    - Uses "intrepid" as the default release.


test.py
# -*- Encoding: UTF-8 -*-
# Copyright (c) 2008 Henri Häkkinen
#
# This file is part of the UbuntuMan Supybot IRC plugin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from supybot.test import *

UbuntuMan = plugin.loadPluginModule('UbuntuMan')
UMConf = conf.supybot.plugins.UbuntuMan

class UbuntuManTestCase(PluginTestCase):
    plugins = ('UbuntuMan',)

    def testMan(self):
        self.assertRegexp('man grep', '^grep.*\|')
        self.assertRegexp('man ls', '^ls.*\|')
        self.assertRegexp('man asdasd', '^No manual page for')
        # test length limit
        m = self.getMsg('man grep')
        self.assertTrue(len(m.args[1]) <= 300)

    def testManurl(self):
        (base, rel, lang) = (UMConf.baseurl, UMConf.release, UMConf.language)
        url = '%s/%s/%s/man1/cat.1.html' % (base, rel, lang)
        self.assertResponse('manurl cat', url)
        url = '%s/%s/%s/man1/cp.1.html' % (base, 'dapper', 'es')
        self.assertResponse('manurl cp --rel dapper --lang es', url)
        self.assertRegexp('manurl asdasd', '^No manual page for')

    def testLanguages(self):
        for s in ('en', 'es', 'de', 'it', 'fr'):
            self.assertNotRegexp('man ls --lang %s' % s, '^Failed to parse')
            self.assertNotRegexp('man bash --lang %s' % s, '^Failed to parse')
        self.assertNotRegexp('man su --lang fi', '^Failed to parse')
        # this fails for now
        #self.assertNotRegexp('man aptitude --lang fi', '^Failed to parse')

    def testFormat(self):
        cmd = 'grep'
        confbak = conf.supybot.plugins.UbuntuMan.format()
        try:
            self.assertNotRegexp('man %s' % cmd, r'\$')
            conf.supybot.plugins.UbuntuMan.format.setValue('$command')
            self.assertResponse('man %s' % cmd, cmd)
            conf.supybot.plugins.UbuntuMan.format.setValue('$url')
            self.assertRegexp('man %s' % cmd,
                    r'^http://manpages\.ubuntu\.com/manpages/\w+/en/.+html')
            conf.supybot.plugins.UbuntuMan.format.setValue('$name')
            self.assertRegexp('man %s' % cmd, r'^%s.{10}' % cmd)
            conf.supybot.plugins.UbuntuMan.format.setValue('$synopsis')
            self.assertRegexp('man %s' % cmd, r'^%s \[OPTIONS\]' % cmd)
            conf.supybot.plugins.UbuntuMan.format.setValue('$description')
            self.assertRegexp('man %s' % cmd, r'^%s.{10}' % cmd)
            conf.supybot.plugins.UbuntuMan.format.setValue(
                   'prefix | $name | $url | $description | $synopsis | subfix')
            self.assertRegexp('man %s' % cmd, r'^prefix.*subfix$')
        finally:
            conf.supybot.plugins.UbuntuMan.format.setValue(confbak)

# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


config.py
# -*- Encoding: UTF-8 -*-
# Copyright (c) 2008 Henri Häkkinen
#
# This file is part of the UbuntuMan Supybot IRC plugin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import supybot.conf as conf
import supybot.registry as registry

def configure(advanced):
    from supybot.questions import expect, anything, something, yn
    UbuntuMan = conf.registerPlugin('UbuntuMan', True)

    if advanced == False:
        return

    baseurl = something("""What value should be used for base URL?""",
                        default="http://manpages.ubuntu.com/manpages")

    release = expect("""What value should be used as the default Ubuntu release?""",
                     possibilities=['dapper', 'gutsy', 'hardy', 'intrepid',
                         'jaunty'],
                     default='intrepid')

    sections = something("""What manual page sections should be enabled?""",
                         default='1 5 8')

    language = expect("""Which language should be used by default?""",
                      possibilities=['en', 'es', 'de', 'fi'],
                      default='en')

    UbuntuMan.baseurl.setValue(baseurl)
    UbuntuMan.release.setValue(release)
    UbuntuMan.sections.setValue(sections)
    UbuntuMan.language.setValue(language)


UbuntuMan = conf.registerPlugin('UbuntuMan')

conf.registerGlobalValue(UbuntuMan, 'baseurl',
    registry.String('http://manpages.ubuntu.com/manpages',
             """Determines the base URL of the manpage repository.
                Do not end this variable to slash."""))

conf.registerGlobalValue(UbuntuMan, 'release',
    registry.String('intrepid',
             """Determines the default release to fetch the manual pages for."""))

conf.registerGlobalValue(UbuntuMan, 'sections',
    registry.SpaceSeparatedListOfStrings(['1', '5', '8'],
             """Determines the list of enabled manual page sections."""))

conf.registerGlobalValue(UbuntuMan, 'language',
    registry.String('en',
             """Determines the default language. Currently supported: de, en,
             es, fi, fr, it"""))

conf.registerGlobalValue(UbuntuMan, 'format',
        registry.String(
    '$name | $synopsis | $description',
    """Determines de format of the reply. Supported keywords: $command,
    $url, $name, $synopsis, $description"""))

conf.registerGlobalValue(UbuntuMan, 'maxLength',
    registry.Integer(300, """Determines maximun length of the output, if
        supybot.reply.mores.length has a value other than zero this register
        has no effect."""))

# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


__init__.py
# -*- Encoding: UTF-8 -*-
# Copyright (c) 2008 Henri Häkkinen
#
# This file is part of the UbuntuMan Supybot IRC plugin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
Provides commands for displaying UNIX manual pages from the Ubuntu Manpage
Repository.
"""

import supybot
import supybot.world as world

__version__ = '1.2.0'

__author__ = supybot.Author('Henri Hakkinen', 'henux', 'henux@users.sourceforge.net')

# This is a dictionary mapping supybot.Author instances to lists of
# contributions.
__contributors__ = {supybot.Author('Terence Simpson', 'stdin', ''): '',
                    supybot.Author('Elián Hanisch', 'm4v', 'lambdae2@gmail.com'): ''}

# This is a url where the most recent plugin package can be downloaded.
__url__ = 'http://henux.nor.fi/coding/projects/9-other/7-ubuntuman'
 # 'http://supybot.com/Members/yourname/UbuntuMan/download'

import config
import plugin
reload(plugin) # In case we're being reloaded.
# Add more reloads here if you add third-party modules and want them to be
# reloaded when this plugin is reloaded.  Don't forget to import them as well!

if world.testing:
    import test

Class = plugin.Class
configure = config.configure


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:


plugin.py
# -*- Encoding: UTF-8 -*-
# Copyright (c) 2008 Henri Häkkinen
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import supybot.log as log
import supybot.conf as conf
import supybot.utils as utils
from supybot.commands import *
import supybot.plugins as plugins
import supybot.ircutils as ircutils
import supybot.registry as registry
import supybot.callbacks as callbacks

import sys

class UbuntuManError(Exception):
    """Ubuntu manual page exception.  Raised when an expected section is
    not found from the manual page or other parsing related error is
    encountered."""
    pass

def cut(s, limit):
    """Function for cutting a string nicely at the end of a word."""
    idx = s[:limit - 3].rfind(' ')
    s = '%s ...' % s[:idx]
    return s

class KeywordsParser:
    """Class for check which sections of the manpage are needed."""
    def __init__(self):
        self.keys = ('description', 'name', 'synopsis')
        self.reset()

    def reset(self):
        self.keysParsed = list()
        for key in self.keys:
            setattr(self, key, False)

    def checkKeywords(self, format):
        L = list()
        for key in self.keys:
            if ('$' + key) in format:
                setattr(self, key, True)
                L.append(key)
            else:
                setattr(self, key, False)
        self.keysParsed = L


class UbuntuManParser:
    """Ubunutu manual page parser."""

    def __init__(self):
        self.keywords = KeywordsParser()
        for key in self.keywords.keys:
            setattr(self, key, '')

    def skipToSection(self, fd, section):
        """Skips lines until '<h3>SECTION</h3>' is found and returns the
        rest of the line as a string with whitespaces normalized and HTML
        tags removed.  'section' can be a string or a tuple of strings."""
        if section.__class__ == str: # check if section is a string
            section = (section, )
        while True:
            ln = fd.readline()
            if not ln:
                raise UbuntuManError('Section %s not found.' % \
                        ', '.join(section))
            for s in section:
                tag = '<h3>%s</h3>' % s
                idx = ln.find(tag)
                if idx > -1:
                    ln = ln[idx + len(tag) + 5:]
                    ln = utils.web.htmlToText(ln, tagReplace='')
                    ln = utils.str.normalizeWhitespace(ln)
                    return ln

    def parseName(self, fd, section='NAME'):
        """Parse the NAME section."""
        #log.debug('UbuntuManParser.parseName() parsing ...')
        self.name = self.skipToSection(fd, section)

    def parseSynopsis(self, fd, section='SYNOPSIS'):
        """Parse the SYNOPSIS section.  Only the first line is read."""
        #log.debug('UbuntuManParser.parseSynopsis() parsing ...')
        self.synopsis = self.skipToSection(fd, section)

    def parseDesc(self, fd, section='DESCRIPTION'):
        """Parse the DESCRIPTION section.  Only the first sentences that fit a
        150 char limit are read."""
        #log.debug('UbuntuManParser.parseDesc() parsing ...')
        description = self.skipToSection(fd, section)
        while len(description) < 300:
            ln = fd.readline()
            if not ln or ln.startswith(' </pre>') or len(ln) == 0:
                break
            if ln.endswith('\xe2\x80\x90\x0a'):
                description += ln[:len(ln) - 4]
            else:
                description = '%s%s ' %(description, ln.strip())
        description = utils.web.htmlToText(description, tagReplace='')
        description = utils.web.normalizeWhitespace(description)
        idx = description[:150].rfind('.')
        if idx < 1:
            description = cut(description, 150)
        else:
            description = description[:idx + 1]
        self.description = description

    def parse(self, fd, command, format):
        """Parse the HTML manual page from the given file descriptor."""
        self.command = command
        self.keywords.checkKeywords(format)
        for x in range(43):
            fd.readline()
        self.keywords.name and self.parseName(fd)
        self.keywords.synopsis and self.parseSynopsis(fd)
        self.keywords.description and self.parseDesc(fd)

class UbuntuManParser_en(UbuntuManParser):
    """Ubuntu manual page parser for English."""
    pass

class UbuntuManParser_es(UbuntuManParser):
    """Ubuntu manual page parser for Spanish."""

    def parseName(self, fd):
        UbuntuManParser.parseName(self,fd, ('NOMBRE', 'NAME'))

    def parseSynopsis(self, fd):
        UbuntuManParser.parseSynopsis(self, fd, ('SINOPSIS', 'SINTAXIS',
            'SYNOPSIS'))

    def parseDesc(self, fd):
        # Should be just DESCRIPCIÓN, but meh :/
        UbuntuManParser.parseDesc(self, fd,
                                  ('DESCRIPCI   N',
                                   'DESCRIPCI?N',
                                   'DESCRIPCION',
                                   'DESCRIPTION',
                                   'DESCRIPCIÓN'))

class UbuntuManParser_de(UbuntuManParser):
    """Ubunutu manual page parser for German."""

    def parseName(self, fd):
        UbuntuManParser.parseName(self, fd, ('BEZEICHNUNG', 'NAME'))

    def parseSynopsis(self, fd):
        # German synopsis sections aren't formated right, taking that into
        # account..
        section = 'BERSICHT'
        while True:
            ln = fd.readline()
            if not ln:
                raise UbuntuManError('Section %s not found.' % section)
            tag = section
            offs = ln.find(tag)
            if offs == -1:
                continue
            ln = fd.readline()
            ln = utils.web.htmlToText(ln, tagReplace='')
            ln = utils.str.normalizeWhitespace(ln)
            break
        self.synopsis = ln

    def parseDesc(self, fd):
        UbuntuManParser.parseDesc(self, fd, section='BESCHREIBUNG')

class UbuntuManParser_fi(UbuntuManParser):
    """Ubuntu manual page parser for Finnish."""

    def parseName(self, fd):
        UbuntuManParser.parseName(self, fd, ('NAME', 'NIMI'))

    def parseSynopsis(self, fd):
        # FIXME aptitude manpage fails
        UbuntuManParser.parseSynopsis(self, fd, ('SYNOPSIS', 'YLEISKATSAUS'))

    def parseDesc(self, fd):
        UbuntuManParser.parseDesc(self, fd, 'KUVAUS')

class UbuntuManParser_it(UbuntuManParser):
    """Ubuntu manual page parser for Italian."""

    def parseName(self, fd):
        UbuntuManParser.parseName(self, fd, ('NOME', 'NAME'))

    def parseSynopsis(self, fd):
        UbuntuManParser.parseSynopsis(self, fd, ('SINTASSI', 'SYNOPSIS'))

    def parseDesc(self, fd):
        UbuntuManParser.parseDesc(self, fd, ('DESCRIZIONE', 'DESCRIPTION'))


class UbuntuManParser_fr(UbuntuManParser):
    """Ubuntu manual page parser for French."""

    def parseName(self, fd):
        UbuntuManParser.parseName(self, fd, ('NOM', 'NAME'))


class UbuntuMan(callbacks.Plugin):
    """This plugin provides commands for displaying UNIX manual pages from
    the Ubuntu Manpage repository."""

    def __init__(self, irc):
        self.__parent = super(UbuntuMan, self)
        self.__parent.__init__(irc)
        self.currentParser = None

    def __setParser(self, language):
        parserClass  = 'UbuntuManParser_' + language
        if parserClass == self.currentParser:
            # Don't initialise the parser again if it isn't needed.
            return
        module = sys.modules['UbuntuMan.plugin']
        # Looks for the parser class that matchs the language set in the
        # config, or defaults to UbuntuManParser_en.
        self.parser = hasattr(module , parserClass) \
            and getattr(module,parserClass)() \
            or UbuntuManParser_en()
        self.currentParser = parserClass

    def __buildUrl(self, release, section, command, language):
        """Build URL to a manual page."""
        if not release:
            release = self.registryValue('release')
        url = '/%s/%s/man%s/%s.%s.html' % (release, language, section,
                    command, section)
        url = self.registryValue('baseurl') + utils.web.urlquote(url)
        return url

    def __tryUrl(self, url):
        """Try to open the given URL.  If succeeds, returns it's file
        descriptor; otherwise returns None."""
        try:
            return utils.web.getUrlFd(url)
        except utils.web.Error:
            return None

    def __getManPageFd(self, release, command, language):
        """Get a file descriptor to the manual page in the Ubuntu Manpage
        Repository."""
        if language == 'en':
            language = (language, )
        else:
            language = (language, 'en')
        for section in self.registryValue('sections'):
            for lang in language:
                url = self.__buildUrl(release, section, command, lang)
                #self.log.debug('UbuntuMan: Trying url %s' % url)
                fd = self.__tryUrl(url)
                if fd:
                    #self.log.debug('UbuntuMan: Success')
                    self.__setParser(lang)
                    self.parser.url = url
                    self.parser.command = command
                    return fd
        return None

    def __formatReply(self):
        """Format the data for the IRC reply."""
        format = self.registryValue('format')
        vars = {
                'url':self.parser.url,
                'command':self.parser.command,
                'name':self.parser.name,
                'synopsis':self.parser.synopsis,
                'description':self.parser.description,
               }
        replace = lambda : utils.str.perlVariableSubstitute(vars, format)
        msg = replace()
        length = conf.supybot.reply.mores.length()
        if not length:
            length = self.registryValue('maxLength')
        if len(msg) > length:
            # if we exceed in length lest try to cut one of the vars
            # without ruining the format.
            for var in self.parser.keywords.keysParsed:
                cutLength = len(msg) - length
                vars[var] = cut(vars[var], - cutLength)
                msg = replace()
                break
            if len(msg) > length:
                # alright, length is really just too short
                msg = '%s ...' %(msg[:length - 4])
        return msg

    def man(self, irc, msg, args, command, optlist):
        """<command> [--rel <release>] [--lang <language>]

        Displays a manual page from the Ubuntu Manpage Repository."""
        release = None
        language = self.registryValue('language')
        format = self.registryValue('format')
        for (opt, arg) in optlist:
            if opt == 'rel':
                release = arg
            elif opt == 'lang':
                language = arg
        try:
            fd = self.__getManPageFd(release, command, language)
            if not fd:
                irc.reply('No manual page for \'%s\'' % command)
                return
            self.parser.parse(fd, command, format)
            #for var in self.parser.keywords.keysParsed:
            #    log.debug('UbuntuMan.man() %s=\'%s\'' % (var,
            #        getattr(self.parser, var)))
            fd.close()
            msg = self.__formatReply()
            irc.reply(msg)
        except UbuntuManError, e:
            irc.reply('Failed to parse the manpage for \'%s\': %s' % (command,
                e.message))
            self.log.info(
                'plugins.UbuntuMan: Failed to parse the manpage in \'%s\'. ' \
                'Report it to the plugin maintainer.' % self.parser.url)

    man = wrap(man, ['something', getopts({'rel':'something',
                                           'lang':'something'})])

    def manurl(self, irc, msg, args, command, optlist):
        """<command> [--rel <release>] [--lang <language>]

        Gives the URL to the full manual page in the Ubuntu Manpage
        Repository."""
        release = None
        language = self.registryValue('language')
        for (opt, arg) in optlist:
            if opt == 'rel':
                release = arg
            elif opt == 'lang':
                language = arg
        try:
            fd = self.__getManPageFd(release, command, language)
            if not fd:
                irc.reply('No manual page for \'%s\'' % command)
                return
            irc.reply(fd.url)
            fd.close()
        except:
            pass

    manurl = wrap(manurl, ['something', getopts({'rel':'something',
                                                 'lang':'something'})])


Class = UbuntuMan


# vim:set shiftwidth=4 tabstop=4 expandtab textwidth=79:
